lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp

```diff
@@ -863,31 +862,91 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
+          }
 
-          if (axesShape.size() != 1 || axesShape[0] == Torch::kUnknownSize)
-            return failure();
+          auto areDistinct = ([](SmallVector<int64_t> array) -> bool {
+            int n = array.size();
+            llvm::SetVector<int64_t> set;
+            for (int i = 0; i < n; i++) {
+              set.insert(array[i]);
+            }
 
-          Value zero = rewriter.create<Torch::ConstantIntOp>(
-              binder.getLoc(), rewriter.getType<Torch::IntType>(),
-              rewriter.getI64IntegerAttr(0));
-          int64_t numAxes = axesShape[0];
-          for (int64_t i = 0; i < numAxes; ++i) {
-            Value iv = rewriter.create<Torch::ConstantIntOp>(
+            // If all elements are distinct, then the size of set should be same
+            // as array's size.
+            return (set.size() == array.size());
+          });
+
+          // If the input shape and result shape is statically known then the
+          // list of dims to be squeezed can be derived from those shapes. As a
+          // result, we don't have to wait for the dim values to be known at
+          // runtime which is also expected by the downstream pipeline.
+          if (inputType.areAllSizesKnown() && resultType.areAllSizesKnown()) {
+            SmallVector<int64_t> inputShape{inputType.getSizes()};
+            SmallVector<int64_t> resultShape{resultType.getSizes()};
+            if (llvm::equal(inputShape, resultShape)) {
+              // Case: none of the dimension is reduced.
+              rewriter.replaceOp(binder.op, data);
+              return success();
+            }
+            if (areDistinct(inputShape)) {
+              // The check for the input shape elements to be distinct is added
+              // for the cases like:
+              // Input: [3, 2, 2] -> Output: [3, 2]
+              // For the above case, from the input and output shape it can't be
+              // inferred whether the dim:1 is reduced or dim:2. To avoid these
+              // type of cases, the check has been placed.
+              SmallVector<int64_t> reduceDims;
+              unsigned resultShapeCounter = 0;
+              for (unsigned i = 0; i < inputShape.size(); i++) {
+                if (resultShapeCounter < resultShape.size() &&
+                    inputShape[i] == resultShape[resultShapeCounter]) {
+                  resultShapeCounter++;
+                } else {
+                  reduceDims.push_back(i);
+                  if (resultShapeCounter < resultShape.size() &&
+                      resultShape[resultShapeCounter] == 1)
+                    resultShapeCounter++;
+                }
+              }
+              for (auto i : reduceDims) {
+                axesList.push_back(rewriter.create<Torch::ConstantIntOp>(
+                    binder.getLoc(), rewriter.getI64IntegerAttr(i)));
+              }
+            }
+          }
+
+          if (axesList.empty()) {
```

Comment:
Ah, to my last comment, I was wrong about this because I thought the axesVal binder check had ended with the brackets before. So we get here if axesVal was indeed provided AND we have an unknown size in either the input or output.

Reply:
Yes, you are correct.